drivers/esp-hosted/esp_hosted_wifi.c

```diff
@@ -0,0 +1,636 @@
+    esp_header->checksum = 0;
+    uint8_t *buf = (uint8_t *)esp_header;
+    for (size_t i = 0; i < (esp_header->len + sizeof(esp_header_t)); i++) {
+        checksum += buf[i];
+    }
+    return checksum;
+}
+
+#if ESP_HOSTED_DEBUG
+static void esp_hosted_dump_header(esp_header_t *esp_header) {
+    static const char *if_strs[] = { "STA", "AP", "SERIAL", "HCI", "PRIV", "TEST" };
+    if (esp_header->if_type > ESP_HOSTED_MAX_IF) {
+        return;
+    }
+    debug_printf("esp header: if %s_IF length %d offset %d checksum %d seq %d flags %x\n",
+        if_strs[esp_header->if_type], esp_header->len, esp_header->offset,
+        esp_header->checksum, esp_header->seq_num, esp_header->flags);
+
+    if (esp_header->if_type == ESP_HOSTED_SERIAL_IF) {
+        tlv_header_t *tlv_header = (tlv_header_t *)(esp_header->payload);
+        debug_printf("tlv header: ep_type %d ep_length %d ep_value %.8s data_type %d data_length %d\n",
+            tlv_header->ep_type, tlv_header->ep_length,
+            tlv_header->ep_value, tlv_header->data_type, tlv_header->data_length);
+    }
+}
+#endif
+
+static int32_t esp_hosted_resp_value(CtrlMsg *ctrl_msg) {
+    // Each response struct return value is located at a different offset,
+    // the following array maps response CtrlMsgs to return values (resp)
+    // offsets within each response struct.
+    const static size_t ctrl_msg_resp_offset[] = {
+        offsetof(CtrlMsgRespGetMacAddress, resp),
+        offsetof(CtrlMsgRespSetMacAddress, resp),
+        offsetof(CtrlMsgRespGetMode, resp),
+        offsetof(CtrlMsgRespSetMode, resp),
+        offsetof(CtrlMsgRespScanResult, resp),
+        offsetof(CtrlMsgRespGetAPConfig, resp),
+        offsetof(CtrlMsgRespConnectAP, resp),
+        offsetof(CtrlMsgRespGetStatus, resp),
+        offsetof(CtrlMsgRespGetSoftAPConfig, resp),
+        offsetof(CtrlMsgRespSetSoftAPVendorSpecificIE, resp),
+        offsetof(CtrlMsgRespStartSoftAP, resp),
+        offsetof(CtrlMsgRespSoftAPConnectedSTA, resp),
+        offsetof(CtrlMsgRespGetStatus, resp),
+        offsetof(CtrlMsgRespSetMode, resp),
+        offsetof(CtrlMsgRespGetMode, resp),
+        offsetof(CtrlMsgRespOTABegin, resp),
+        offsetof(CtrlMsgRespOTAWrite, resp),
+        offsetof(CtrlMsgRespOTAEnd, resp),
+        offsetof(CtrlMsgRespSetWifiMaxTxPower, resp),
+        offsetof(CtrlMsgRespGetWifiCurrTxPower, resp),
+        offsetof(CtrlMsgRespConfigHeartbeat, resp),
+    };
+
+    int32_t resp = -1;
+    size_t index = ctrl_msg->msg_id - CTRL_MSG_ID__Resp_Base;
+
+    // All types of messages share the same payload base address.
+    if (ctrl_msg->resp_get_mac_address != NULL &&
+        ctrl_msg->msg_type == CTRL_MSG_TYPE__Resp &&
+        index > 0 && index <= MP_ARRAY_SIZE(ctrl_msg_resp_offset)) {
+        // Return the response struct's return value.
+        size_t offset = ctrl_msg_resp_offset[index - 1];
+        resp = *((int32_t *)((char *)ctrl_msg->resp_get_mac_address + offset));
```

Comment:
> This is not magic, the common base address of union members is well-defined by the C standard and offsetof is standard, also this driver, the ESP one and protobuf itself all function using the same logic more or less.

I don't mean it's non-standard, I mean it relies on the implementation of a code generation layer that Micropython doesn't control (Espressif have control over both sides of this interface, we only have control on this side.)

The protobuf compiler can generate code like this because it's generating it from the .proto file at compile time, so if the proto file changes then the code generation changes automatically. There's no code generation tool here to pick up such a change, if it happens then a programmer has to notice it and manually make the change.

It's pretty unlikely that a breaking change happens that messes up the order of the numeric values, and that the programmer who updates this driver doesn't notice it. However I've been burned enough times by my own pointer tricks hiding these kind of weird bugs to question them intensely these days. :)

That said, if you're confident with this then of course it's fine to keep. It's simple enough to change in the future if the protocol changes.